components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/AssignUserRolesFunctionImpl.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.CarbonException;
+import org.wso2.carbon.core.util.AnonymousSessionUtil;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsAuthenticatedUser;
+import org.wso2.carbon.identity.application.authentication.framework.exception.FrameworkException;
+import org.wso2.carbon.identity.conditional.auth.functions.user.internal.UserFunctionsServiceHolder;
+import org.wso2.carbon.user.core.UserRealm;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+
+import java.util.List;
+
+/**
+ * Function to update given roles for a given user.
+ * The purpose is to perform role assigning during dynamic authentication.
+ */
+public class AssignUserRolesFunctionImpl implements AssignUserRolesFunction {
+
+    private static final Log LOG = LogFactory.getLog(AssignUserRolesFunctionImpl.class);
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param user           Authenticated user.
+     * @param assigningRoles Roles to be assigned.
+     * @return <code>true</code> If the role assigning is successfully completed. <code>false</code> for any other case.
+     */
+    @Override
+    public boolean assignUserRoles(JsAuthenticatedUser user, List<String> assigningRoles) {
+
+        if (user != null && assigningRoles != null) {
+            try {
+                if (user.getWrapped() != null) {
+                    String tenantDomain = user.getWrapped().getTenantDomain();
+                    String userStoreDomain = user.getWrapped().getUserStoreDomain();
+                    String username = user.getWrapped().getUserName();
+                    UserRealm userRealm = getUserRealm(tenantDomain);
+                    if (userRealm != null) {
+                        UserStoreManager userStore = getUserStoreManager(tenantDomain, userRealm, userStoreDomain);
+                        if (userStore != null) {
+                            userStore.updateRoleListOfUser(
+                                    username,
+                                    new String[0],
+                                    assigningRoles.toArray(new String[0])
+                            );
+                            return true;
+                        } else {
+                            LOG.debug("Null value user store received for the user "
+                                    + user.getMember("username"));
+                        }
+                    } else {
+                        LOG.debug("Null value user realm received for the user "
+                                + user.getMember("username"));
+                    }
+                } else {
+                    LOG.debug("Null value received while getting wrapped content for user");
+                }
+            } catch (UserStoreException e) {
+                LOG.error("Error in getting user from store at the function ", e);
```

Comment:
what is meant by "at the function"?
We can make it like "Error while getting user: " + username + " from user store .."